### PR TITLE
Added pad test, update LegalizeInputTypes to allow linalg.yield

### PR DESCRIPTION
**Pull request by rsuderman**
LegalizeInputTypes assumes no yield ops, updated so that it can handle nested
linalg blocks.

Simple pad test to verify.

### CODE REVIEW

File: iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp [hunk 1]
```diff
@@ -85,7 +85,8 @@ LogicalResult convertOperation(Operation *oldOp,
                                FlowTypeConverter &typeConverter,
                                BlockAndValueMapping &mapping,
                                OpBuilder &builder) {
-  if (llvm::isa<mlir::linalg::LinalgOp>(oldOp)) {
+  if (llvm::isa<mlir::linalg::LinalgOp>(oldOp) ||
+      llvm::isa<mlir::linalg::YieldOp>(oldOp)) {
     // Currently assumes linalg ops have legal types.
     // TODO: rewrite to generic and back.
```
Comment (ScottTodd) [inline iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp:91]: I don't really understand the TODO here, and you're adding another case to the condition that it comments on. Can you expand on the TODO?

This now also assumes that YieldOps are already legal? What would "rewriting to generic and back" accomplish?

File: iree/test/e2e/tosa_ops/pad.mlir [hunk 2]
```diff
@@ -0,0 +1,8 @@
+
+func @pad_test() attributes { iree.module.export } {
+    %0 = "tosa.const"() { value = dense<1> : tensor<1x1x1xi32> } : ()  -> (tensor<1x1x1xi32>)
+    %1 = "tosa.const"() { value = dense<[[0, 1], [1, 0], [0, 0]]> : tensor<3x2xi32> } : ()  -> (tensor<3x2xi32>)
+    %result = "tosa.pad"(%0, %1)  : (tensor<1x1x1xi32>, tensor<3x2xi32>)  -> (tensor<2x2x1xi32>)
+    check.expect_eq_const(%result, dense<[[[0], [1]], [[0], [0]]]> : tensor<2x2x1xi32>) : tensor<2x2x1xi32>
```
Comment (ScottTodd) [inline iree/test/e2e/tosa_ops/pad.mlir:6]: o_O seems to match the spec, but wow the spec is confusing... 

After reading through these:
https://mlir.llvm.org/docs/Dialects/TOSA/#tosapad-mlirtosapadop
https://developer.mlplatform.org/w/tosa/
and referencing this test, I'm still not sure where `2x2x1` is coming from, or why these values are expected...

Could you use values other than 0 and 1 to more clearly show what is happening?

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: iree/test/e2e/tosa_ops/pad.mlir (hunk 2 above)
Comment (rsuderman) [reply iree/test/e2e/tosa_ops/pad.mlir]: I added two more tests and used the internal value as 42 to be more clear. I can not change what the padding value is, so you are stuck with the 0s. Not sure we are going to get clearer with padding.

File: iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp (hunk 1 above)
Comment (rsuderman) [reply iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp]: I updated to check directly that we assume all linalg ops are valid (I think this is what the original code was meant to do). Tried to update the comment but the original line was a bit of a mystery to me too.

**Review by ScottTodd — APPROVED** [review]

File: iree/test/e2e/tosa_ops/pad.mlir [hunk 3]
```diff
@@ -0,0 +1,23 @@
+func @pad_1D_test() attributes { iree.module.export } {
+    %0 = "tosa.const"() { value = dense<42> : tensor<2xi32> } : ()  -> (tensor<2xi32>)
+    %1 = "tosa.const"() { value = dense<[[3, 2]]> : tensor<1x2xi32> } : ()  -> (tensor<1x2xi32>)
+    %result = "tosa.pad"(%0, %1)  : (tensor<2xi32>, tensor<1x2xi32>)  -> (tensor<7xi32>)
```
Comment (hanhanW) [inline iree/test/e2e/tosa_ops/pad.mlir:4]: Do we test with "tosa.const"? If they are constants, I think they should/will be folded...

Comment (rsuderman) [reply iree/test/e2e/tosa_ops/pad.mlir]: Good catch. We actually don't have folders in TOSA so it technically works find but better to make input-1 unfoldable.

File: iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp [hunk 4]
```diff
@@ -85,9 +85,9 @@ LogicalResult convertOperation(Operation *oldOp,
...
                                OpBuilder &builder) {
-  if (llvm::isa<mlir::linalg::LinalgOp>(oldOp)) {
-    // Currently assumes linalg ops have legal types.
-    // TODO: rewrite to generic and back.
+  if (isa<linalg::LinalgDialect>(oldOp->getDialect())) {
+    // Currently we assume all Linalg ops only contain valid types.
+    // TODO: Support converting Linalg types when unsupported.
     builder.clone(*oldOp, mapping);
     return success();
   }
```
Comment (hanhanW) [inline iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp:93]: I personally think we should have a test for this in legalize_input_types.mlir... It looks like you are trying to allow linalg.yield, but it should only exist in Linalg bodys?

Comment (rsuderman) [reply iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp:93]: Correct however this pass recurses into linalg bodies (probably shouldn't but it does). There are also other linalg ops that will likely appear so we should mark all of linalg as a special case until it is resolved. Test added.

**Pull request merged**